src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java

```diff
@@ -728,57 +709,59 @@ private boolean createFeedbackForDuplicateTests(Result result, ProgrammingExerci
+        double score = 0D;
+
+        if (!hasDuplicateTestCases) {
+            score = calculateScore(programmingExercise, allTestCases, result, staticCodeAnalysisFeedbacks, applySubmissionPolicy);
         }
 
-        result.getFeedbacks().forEach(feedback -> {
-            if (feedback.getCredits() == null) {
-                feedback.setCredits(0D);
-            }
+        result.setScore(score, programmingExercise.getCourseViaExerciseGroupOrCourseMember());
+    }
+
+    private void updateFeedbackCredits(Result result, Set<ProgrammingExerciseTestCase> allTestCases, Double exerciseMaxPoints) {
+        // Set credits for successful test cases
+        allTestCases.stream().filter(testCase -> testCase.isSuccessful(result)).forEach(testCase -> {
+            double credits = calculatePointsForTestCase(result.getParticipation(), exerciseMaxPoints, testCase, allTestCases);
+            setCreditsForTestCaseFeedback(result, testCase, credits);
         });
+
+        result.getFeedbacks().stream().filter(feedback -> Objects.isNull(feedback.getCredits())).forEach(feedback -> feedback.setCredits(0D));
     }
 
     /**
      * Calculates the score of automatic test cases for the given result with possible penalties applied.
-     * @param programmingExercise the result belongs to.
-     * @param allTests that should be considered in the score calculation.
-     * @param result for which a score should be calculated.
-     * @param successfulTestCases all test cases that passed for the submission.
+     *
+     * @param programmingExercise        the result belongs to.
+     * @param allTests                   that should be considered in the score calculation.
+     * @param result                     for which a score should be calculated.
      * @param staticCodeAnalysisFeedback that has been created for the submission.
-     * @param applySubmissionPolicy true, if penalties from submission policies should be applied.
-     * @return the final total score that should be given to the result.
+     * @param applySubmissionPolicy      true, if penalties from submission policies should be applied.
+     * @return the final total score in percent that should be given to the result.
      */
     private double calculateScore(final ProgrammingExercise programmingExercise, final Set<ProgrammingExerciseTestCase> allTests, final Result result,
-            final Set<ProgrammingExerciseTestCase> successfulTestCases, final List<Feedback> staticCodeAnalysisFeedback, boolean applySubmissionPolicy) {
-        if (successfulTestCases.isEmpty()) {
+            final List<Feedback> staticCodeAnalysisFeedback, boolean applySubmissionPolicy) {
+        // TODO: this might be redundant and covered in calculateSuccessfulTestPoints
+        var noSuccessfulTestCases = allTests.stream().filter(testCase -> testCase.isSuccessful(result)).toList().isEmpty();
```

Comment:
```suggestion
        boolean noSuccessfulTestCases = allTests.stream().allMatch(testCase -> !testCase.isSuccessful(result));
```
two things:
- guidelines: no `var` for simple types
- no need to construct the list, that way the operation can short-circuit when one non-matching item in the stream is found